src/Combinatorics/Matroids/matroids.jl

```diff
@@ -0,0 +1,769 @@
+julia> M = matroid_from_bases([[1,2],[1,'i'],[1,'j'],[2,'i'],[2,'j']],[1,2,'i','j']);
+
+julia> N = deletion(M,'i')
+Matroid of rank 2 on 3 elements
+```
+
+# Example
+```jldoctest
+julia> M = matroid_from_bases([[1,2],[1,'i'],[1,'j'],[2,'i'],[2,'j']],[1,2,'i','j']);
+
+julia> N = deletion(M,['i','j'])
+Matroid of rank 2 on 2 elements
+
+julia> groundset(N)
+2-element Vector{Any}:
+ 1
+ 2
+```
+"""
+function deletion(M::Matroid,set::Union{AbstractVector, Set})
+    sort_set = Vector(undef,length(M.groundset)-length(set))
+    gs2num = Dict{Any,IntegerUnion}()
+    i = 1
+    for elem in M.groundset
+        if length(findall(x->x==elem, set))==0
+            sort_set[i]=elem
+            gs2num[elem] = i
+            i+=1
+        end
+    end
+    pm_del = Polymake.matroid.deletion(M.pm_matroid, Set([M.gs2num[i]-1 for i in set]))
+    return Matroid(pm_del, sort_set, gs2num)
+end
+
+deletion(M::Matroid,elem::Union{IntegerUnion,Char,String}) = deletion(M,Vector([elem]))
+
+@doc Markdown.doc"""
+The `restriction M|S` on a subset `S` of the ground set `E` of the matroid `M`.
+
+See Section 3 of Oxl11 (@cite)
+
+# Example
+```jldoctest
+julia> M = matroid_from_bases([[1,2],[1,'i'],[1,'j'],[2,'i'],[2,'j']],[1,2,'i','j']);
+
+julia> N = restriction(M,[1,2])
+Matroid of rank 2 on 2 elements
+
+julia> groundset(N)
+2-element Vector{Int64}:
+ 1
+ 2
+```
+"""
+function restriction(M::Matroid,set::Union{AbstractVector, Set})
+    sort_set = copy(set)
+    gs2num = Dict{Any,IntegerUnion}()
+    i = 1
+    for elem in M.groundset
+        if length(findall(x->x==elem, set))>0
+            sort_set[i]=elem
+            gs2num[elem] = i
+            i+=1
+        end
+    end
+    pm_complement = setdiff(Set(0:length(M.groundset)-1),Set([M.gs2num[i]-1 for i in set]))
+    pm_rest = Polymake.matroid.deletion(M.pm_matroid, pm_complement)
+    return Matroid(pm_rest, sort_set, gs2num)
+end
+
+@doc Markdown.doc"""
+The `contraction M/S` of an element or a subset `S` of the ground set `E` of the matroid `M`.
+
+See Section 3 of Oxl11 (@cite)
+
+# Example
+```jldoctest
+julia> M = matroid_from_bases([[1,2],[1,'i'],[1,'j'],[2,'i'],[2,'j']],[1,2,'i','j']);
+
+julia> N = contraction(M,'i')
+Matroid of rank 1 on 3 elements
+```
+
+# Example
+```jldoctest
+julia> M = matroid_from_bases([[1,2],[1,'i'],[1,'j'],[2,'i'],[2,'j']],[1,2,'i','j']);
+
+julia> N = deletion(M,['i','j'])
+Matroid of rank 2 on 2 elements
+
+julia> groundset(N)
+2-element Vector{Any}:
+ 1
+ 2
+```
+"""
+function contraction(M::Matroid,set::Union{AbstractVector, Set})
+    sort_set = Vector(undef,length(M.groundset)-length(set))
+    gs2num = Dict{Any,IntegerUnion}()
+    i = 1
+    for elem in M.groundset
+        if length(findall(x->x==elem, set))==0
+            sort_set[i]=elem
+            gs2num[elem] = i
+            i+=1
+        end
+    end
+    pm_contr = Polymake.matroid.contraction(M.pm_matroid, Set([M.gs2num[i]-1 for i in set]))
+    return Matroid(pm_contr, sort_set, gs2num)
+end
```

Comment:
This block is almost identical to the previous `restriction`. It should be refactored to eliminate duplication.

Reply:
And `deletion`.

Reply:
Done.